### PR TITLE
feat: Implement 429 handling for request transport

### PR DESCRIPTION
**Pull request by kamilogorek**
1. Is this a good way to keep state?
2. How to make sure that `Agent` is always initialized in the module? L183 looks dubious.
3. How can I change server response based on the request number? Peeking at req body is meh.
4. Is there an easy way to mock time? Otherwise, 2 commented tests are really hard to implement without passing some time function around.

image

### CODE REVIEW

File: lib/sentry/client.ex [hunk 1]
```diff
@@ -186,18 +213,31 @@ defmodule Sentry.Client do
       Config.hackney_opts()
       |> Keyword.put_new(:pool, @hackney_pool_name)
 
-    with {:ok, 200, _, client} <- :hackney.request(:post, url, headers, body, hackney_opts),
-         {:ok, body} <- :hackney.body(client),
-         {:ok, json} <- json_library.decode(body) do
-      {:ok, Map.get(json, "id")}
-    else
+    case :hackney.request(:post, url, headers, body, hackney_opts) do
```
Comment (kamilogorek) [inline lib/sentry/client.ex]: That's definitely not written with an optimal syntax

**Comment by mitchellhenke** [comment]
@kamilogorek thanks for the pull request and sorry for the delay, I will be taking a look this week!

**Comment by kamilogorek** [comment]
No rush @mitchellhenke , thanks!

File: lib/sentry/client.ex [hunk 2]
```diff
@@ -1,3 +1,23 @@
+defmodule ClientSupervisor do
+  use Agent
+
+  def start_link do
+    Agent.start_link(fn -> Timex.now() end, name: __MODULE__)
```
Comment (mitchellhenke) [inline lib/sentry/client.ex:5]: Can replace with: `NaiveDateTime.utc_now()`

File: lib/sentry/client.ex [hunk 3]
```diff
@@ -387,6 +439,21 @@ defmodule Sentry.Client do
...
 
+  defp parse_retry_after(header) do
+    case Timex.parse(header, "{RFC1123}") do
+      {:ok, retry_after} ->
+        retry_after
+
+      {:error, _} ->
+        try do
+          {retry_after, _} = Integer.parse(header, 10)
+          Timex.shift(Timex.now(), seconds: retry_after)
```
Comment (mitchellhenke) [inline lib/sentry/client.ex:450]: Can replace with:

```elixir
NaiveDateTime.add(NaiveDateTime.utc_now(), retry_after)
```

File: lib/sentry/client.ex [hunk 4]
```diff
@@ -1,3 +1,23 @@
+defmodule ClientSupervisor do
+  use Agent
+
+  def start_link do
+    Agent.start_link(fn -> Timex.now() end, name: __MODULE__)
+  end
+
+  def should_drop_event do
+    Agent.get(__MODULE__, fn state -> Timex.before?(Timex.now(), state) end)
```
Comment (mitchellhenke) [inline lib/sentry/client.ex:9]: Can replace with:
```NaiveDateTime.compare(NaiveDateTime.utc_now(), state) != :lt```

**Comment by kamilogorek** [comment]
Thanks for a review! Updated code for agent and removed timex dependency (except for parsing). Unfortunately, I don't have too much spare time to work on this PR in near future so feel free to take it over if you're able to. Thanks!

**Comment by mitchellhenke** [comment]
@kamilogorek thanks, I don't mind taking it from here!

**Comment by kamilogorek** [comment]
Cleaning up my stale PRs. Feel free to just use branch directly. Thanks :)

**Comment by mitchellhenke** [comment]
@kamilogorek thank you, I plan to get to this soon!

**Pull request closed without merging**